rose/server/score.py

```diff
@@ -27,54 +27,64 @@ def process(players, track):
             if player.x < config.matrix_width - 1:
                 player.x += 1
 
-    # Now handle obstacles, preferring players in their own lane.
+        obstacle = track.get(player.x, player.y)
 
-    sorted_players = sorted(six.itervalues(players),
-                            key=lambda p: 0 if p.in_lane() else 1)
-    positions = set()
+        if obstacle == obstacles.NONE:
+            # Move forlward
+            player.score += config.score_move_forward
 
-    for player in sorted_players:
-        player.score += config.score_move_forward
-        obstacle = track.get(player.x, player.y)
-        if obstacle == obstacles.CRACK:
-            if player.action != actions.JUMP:
-                track.clear(player.x, player.y)
-                player.y += 1
-                player.score += config.score_move_backward * 2
-            else:
-                player.score += config.score_jump
         elif obstacle in (obstacles.TRASH,
                           obstacles.BIKE,
                           obstacles.BARRIER):
-            if player.action not in (actions.LEFT, actions.RIGHT):
+            # Move back
+            track.clear(player.x, player.y)
+            player.y += 1
+            player.score += config.score_move_backward
+
+        elif obstacle == obstacles.CRACK:
+            if player.action == actions.JUMP:
+                # Move forlward
+                player.score += config.score_move_forward
+                player.score += config.score_jump
+            else:
+                # Move back
                 track.clear(player.x, player.y)
                 player.y += 1
-                player.score += config.score_move_backward * 2
+                player.score += config.score_move_backward
+
         elif obstacle == obstacles.WATER:
-            if player.action != actions.BRAKE:
+            if player.action == actions.BRAKE:
+                # Move forlward
+                player.score += config.score_move_forward
+                player.score += config.score_brake
+            else:
+                # Move back
                 track.clear(player.x, player.y)
                 player.y += 1
-                player.score += config.score_move_backward * 2
-            else:
-                player.score += config.score_brake
+                player.score += config.score_move_backward
+
         elif obstacle == obstacles.PENGUIN:
             if player.action == actions.PICKUP:
+                # Move forlward
                 track.clear(player.x, player.y)
                 player.score += config.score_move_forward
+                player.score += config.score_pickup
+            else:
+                # Move forlward no reward
+                player.score += config.score_move_forward
 
-        # Here we can end the game when player gets out of
-        # the track bounds. For now, just keep the player at the same
-        # location.
-        player.y = min(config.matrix_height - 1, max(2, player.y))
```

Comment:
You moved this later, but should be done before we handle collisions. If a player was move back out of the track by hitting an obstacle, we want to move it back to the track. We can do this at the point we change player.y, maybe:

    player.move_backward()

This can do the checks for out-of-track so we don't need to check in the processing loop.

Which is much more clear than:

    player.y += 1

(Working with the inverted coordinates is horrible)

Reply:
makes sense to me too, moved back to before the collision check